### ISSUE TITLE
v26.1.0速度提升数倍，但是训练过程梯度易爆炸

### ISSUE THREAD

**Issue by janelu9** [flashserve/flash-linear-attention-npu#105]
verl的官方镜像：quay.io/ascend/verl   verl-8.5.2-910b-ubuntu22.04-py3.11-qwen3-5 
基础镜像 | Ubuntu 22.04
-- | --
npu | 910b
Python | 3.11
CANN | 8.5.2
torch | 2.9.0
torch_npu | 2.9.0.post3+git66c81d5
triton-ascend | 3.2.0

使用方式如下：
```

          query = query.reshape(batch_size, seq_len, -1, self.head_k_dim)
          key = key.reshape(batch_size, seq_len, -1, self.head_k_dim)
          value = value.reshape(batch_size, seq_len, -1, self.head_v_dim)
          beta = b.sigmoid()
          g = -self.A_log.float().exp() * F.softplus(a.float() + self.dt_bias)
  
          ratio = self.num_local_v_heads // self.num_local_k_heads
          if ratio > 1:
              query = query.repeat_interleave(ratio, dim=2)
              key   = key.repeat_interleave(ratio, dim=2)
              from .flash_gated_delta_rule import flash_gated_delta_rule
              self.chunk_gated_delta_rule = flash_gated_delta_rule

            
            query_npu = pad_4d_h(query,self.max_len).transpose(1, 2).contiguous()
            key_npu   = pad_4d_h(key,self.max_len).transpose(1, 2).contiguous()
            value_npu = pad_4d_h(value,self.max_len).transpose(1, 2).contiguous()
            beta = pad_hidden_states(beta,self.max_len)
            g = pad_hidden_states(g,self.max_len)

            core_attn_out, _ = self.chunk_gated_delta_rule(
                query_npu,
                key_npu,
                value_npu,
                g=g,
                beta=beta,
                output_final_state=False,
                use_qk_l2norm_in_kernel=True,
                cu_seqlens=None, 
            )
            core_attn_out = core_attn_out[:,:seq_len]
```

前两轮训练的还挺正常，中间突然出现loss nan 咋么回事？毫无征兆的梯度爆炸，如果不使用这个接口，全部用torch实现是没有问题的，一切正常的，但是速度实在难以忍受。

**Comment by weinachuan** [comment]
从现象看，优先怀疑不是 fla_npu v26.1.0 算子本身的数值逻辑问题，而是当前环境里的 torch_npu 包没有包含 GDN 相关修复。

你现在使用的是 quay.io/ascend/verl:verl-9.0.0-910b-ubuntu22.04-py3.11-latest 镜像自带的 torch_npu 2.9.0。GDN 这条路径依赖 torch_npu 里的自定义算子适配和 stream 同步逻辑；如果使用的是未包含 GDN 修复的 torch_npu 包，可能出现算子间执行或数据同步没有按预期生效，表现就会比较像“前几轮正常，训练中途突然 NaN / 梯度爆炸”。

建议先按下面顺序排查：

1. 使用 Ascend PyTorch release 页面里提供的、明确包含 GDN 修复的 torch_npu 包：
   https://gitcode.com/Ascend/pytorch/releases?presetConfig={%22tags%22:229,%22release%22:110}

2. 确认 torch / torch_npu / CANN / Python / 架构版本是匹配的一组，不要只看版本号同为 2.9.0 就认为等价。镜像内置包、PyPI 包和 release 页面提供的修复包可能不是同一套内容。

3. 替换 torch_npu 后，建议重新安装或重新编译 fla_npu 相关扩展，避免旧扩展仍然链接到旧的 torch_npu 环境。

4. 复测前可以先打印实际加载路径确认没有用错包：

    import torch
    import torch_npu
    print("torch", torch.__version__)
    print("torch_npu", torch_npu.__version__)
    print("torch_npu path", torch_npu.__file__)

如果换成上述修复包并重新安装 fla_npu 后仍然复现 NaN，再麻烦补充这几项信息，我们继续往算子输入范围或具体 shape 上定位：

- 实际加载的 torch / torch_npu 版本和 torch_npu.__file__
- query/key/value/g/beta 的 dtype、shape、是否包含 NaN/Inf
- 复现时的 use_qk_l2norm_in_kernel、output_final_state、cu_seqlens 配置
- NaN 首次出现是在 forward 输出、loss，还是 backward 梯度里

**Comment by janelu9** [comment]
> 从现象看，优先怀疑不是 fla_npu v26.1.0 算子本身的数值逻辑问题，而是当前环境里的 torch_npu 包没有包含 GDN 相关修复。
> 
> 你现在使用的是 quay.io/ascend/verl:verl-9.0.0-910b-ubuntu22.04-py3.11-latest 镜像自带的 torch_npu 2.9.0。GDN 这条路径依赖 torch_npu 里的自定义算子适配和 stream 同步逻辑；如果使用的是未包含 GDN 修复的 torch_npu 包，可能出现算子间执行或数据同步没有按预期生效，表现就会比较像“前几轮正常，训练中途突然 NaN / 梯度爆炸”。
> 
> 建议先按下面顺序排查：
> 
> 1. 使用 Ascend PyTorch release 页面里提供的、明确包含 GDN 修复的 torch_npu 包：
>    [https://gitcode.com/Ascend/pytorch/releases?presetConfig={%22tags%22:229,%22release%22:110}](https://gitcode.com/Ascend/pytorch/releases?presetConfig=%7B%22tags%22:229,%22release%22:110%7D)
> 2. 确认 torch / torch_npu / CANN / Python / 架构版本是匹配的一组，不要只看版本号同为 2.9.0 就认为等价。镜像内置包、PyPI 包和 release 页面提供的修复包可能不是同一套内容。
> 3. 替换 torch_npu 后，建议重新安装或重新编译 fla_npu 相关扩展，避免旧扩展仍然链接到旧的 torch_npu 环境。
> 4. 复测前可以先打印实际加载路径确认没有用错包：
>    import torch
>    import torch_npu
>    print("torch", torch.**version**)
>    print("torch_npu", torch_npu.**version**)
>    print("torch_npu path", torch_npu.**file**)
> 
> 如果换成上述修复包并重新安装 fla_npu 后仍然复现 NaN，再麻烦补充这几项信息，我们继续往算子输入范围或具体 shape 上定位：
> 
> * 实际加载的 torch / torch_npu 版本和 torch_npu.**file**
> * query/key/value/g/beta 的 dtype、shape、是否包含 NaN/Inf
> * 复现时的 use_qk_l2norm_in_kernel、output_final_state、cu_seqlens 配置
> * NaN 首次出现是在 forward 输出、loss，还是 backward 梯度里

torch==2.9.0, 能否提供一个配套cann8.5.2的 包含 GDN 修复的 torch_npu 包？
